### ISSUE TITLE
Serenity boot protocol

### ISSUE THREAD

**Issue by supercomputer7** [SerenityOS/serenity#552]
We need a custom boot protocol. Benefits of that feature is that we can actually get rid of the complex multiboot scheme as a start. We can keep a support for multiboot so people can still use GRUB, but also add our own boot protocol.

As I talked with @Quaker762 about it, we probably need 4 major things in such boot protocol:
1. Framebuffer data, since Serenity will depend on such data later for actually using the display.
2. Memory Map. Unless Serenity is going to run on a toaster, the boot loader must supply such map. A custom standard form must be established to make sure that we are not going to mess around with e820 conflicting UEFI memory map...
3. RSDP pointer. This is important for ACPI support.
4. Command line data. At some point Serenity will grow enough so it will need such feature. The boot loader simply passes whatever configuration the user has set up. Since it is basically a string, many parameters can be added when needed. My suggestions for parameters are mem_limit (to put a limit on memory counting), boot device (preferably with UUID, so Serenity will know which partition to mount and use), loglevel and such.

If you find that I forgot something please be sure to let me know about it.

A simple structure of the Serenity boot protocol can be like that (this is only a sample for now):

```
typedef struct framebuffer
{
    uint8_t type;
    void *framebuffer;
    uint16_t height;
    uint16_t width;
    uint8_t depth;
} framebuffer_data;

typedef struct mem_unit{
    uint8_t type;
    uint64_t base;
    uint64_t length;
} mem_unit;

typedef struct mem_map
{
    uint32_t length;
    uint32_t usable_mem;
    uint32_t unusable_mem;
    mem_unit records[];
} mem_map;

typedef struct command
{
    uint16_t length;
    char command[];
} command_data;

typedef struct bootdata {
    framebuffer_data fb_data;
    mem_map* memory;
    void *rsdp;
    command_data cmd_data;
} BOOT_DATA;
```
Such support will be implemented in my bootloader:
https://github.com/supercomputer7/nahmanboot
I intend to write a userspace program to ease the installation on a PC system. If I find time, I can try porting such tool to Serenity.

We need to discuss all of this matter more precisely of course.

**Comment by awesomekling** [comment]
Multiboot already provides these features, and is supported by existing bootloaders.

I don't understand why we would need a custom boot protocol.

**Comment by supercomputer7** [comment]
> Multiboot already provides these features, and is supported by existing bootloaders.
> 
> I don't understand why we would need a custom boot protocol.

Well, Multiboot doesn't give you the pointer to RSDP. Multiboot2 however does give you RSDP pointer, but from what I see many OSDevers avoid it since it is way more complex so they prefer building a custom boot protocol. Also, a boot loader can actually take care of the paging stuff if needed.

**Comment by awesomekling** [comment]
> Well, Multiboot doesn't give you the pointer to RSDP. Multiboot2 however does give you RSDP pointer, but from what I see many OSDevers avoid it since it is way more complex so they prefer building a custom boot protocol.

Switching to Multiboot2 looks trivial compared to implementing a custom boot protocol. What is the problem, specifically?

> Also, a boot loader with custom boot protocol can actually take care of the paging stuff if needed.

What do you mean by "take care of the paging stuff"?

**Comment by supercomputer7** [comment]
> > Well, Multiboot doesn't give you the pointer to RSDP. Multiboot2 however does give you RSDP pointer, but from what I see many OSDevers avoid it since it is way more complex so they prefer building a custom boot protocol.
> 
> Switching to Multiboot2 looks trivial compared to implementing a custom boot protocol. What is the problem, specifically?
> 
> > Also, a boot loader with custom boot protocol can actually take care of the paging stuff if needed.
> 
> What do you mean by "take care of the paging stuff"?

One problem I already mentioned is the lack of requirement of boot modules support. The spec (Multiboot & Multiboot2) says the bootloader might allow such feature but doesn't need to implement strictly, so we might encounter problems if we need to load a ramdisk in the future.
Second problem is that the Multiboot 2 is complex and bloated in my opinion. Look at the obsolete "BIOS Boot device" tag in Multiboot2 spec. Instead of letting the bootloader to supply boot device via the commandline, they add useless tags that you must check. No one uses APM anymore but apparently Multiboot2 does support such tags. The spec specify two tags for handling memory map, one for e820 map, the second for EFI memory map, instead of building just one memory map standard for booting. The complexity isn't justified here, they could easily do things less complex.
My suggestion is to build less complex boot protocol, less bloated, more flexible and unique to this project. The code I included is a good example of a simple and easy to understand structure of boot protocol, in my opinion of course.

About the paging - A bootloader that is implementing multiboot & multiboot2 and boot kernels via these methods, must disable paging if it is enabled. If by chance the OS is 64bit and is using the UEFI for booting, paging is actually enabled (by default, blame the UEFI spec) but everything is identity-paged so everything is 1:1 (virtual:physical).
One can change such mess by asking the bootloader to relocate the kernel to a specified physical address then to enable paging with specified instructions on how to install the paging tables and fill them as requested. That is an optional suggestion of course, and some might say that it is an overkill for a bootloader, or this isn't the job of the bootloader. I took the idea mainly from here:
https://wiki.osdev.org/Bootloader_FAQ#What_should_my_bootloader_support.3F
